Read only

joshmoore opened the pull request:
Example of having a read-only user created for a given database. This is for use with read-only OMERO (cF. https://github.com/openmicroscopy/openmicroscopy/pull/5213 etc)

The permissions can only be set _after_ the database has been created. If this is to be part of this or another role, then either table creation will need to take place here or a second role will need to be created.

joshmoore committed:
Use molecule-dependencies

joshmoore committed:
Add read_only example

This configures a user for read only operation
(SELECT only on tables with some tables excluded).

joshmoore committed:
Fix up comments referencing 'roles/'

joshmoore committed:
Add modify test

joshmoore committed:
Remove 'ON CONFLICT' to support PG 9.4

joshmoore committed:
Add test_create.py

manics commented:
Maybe a second role would be easier, especially if omero-read-only becomes widespread? Either a role which operates on existing databases, or if you prefer a postgres modification role specific to omero-read-only?

joshmoore committed:
Omit idempotence test

Unfortunately there doesn't seem to be a way to say
"permit all tables except X" in an idempotent way.
Since `password` is not granted, the ALL_ON_SCHEMA
runs, and then `password` has to be re-ungranted each
time.

joshmoore committed:
Also add 'syntax' step

manics commented:
The danger with removing the `idempotence` step is if in future someone makes a core change that isn't idempotent you get a cascading effect where other roles and playbooks that use this role also fail the idempotence check.

manics commented:
Since this is effectively an example (there aren't any changes to the role) how about one of:
- move this to a development role
- move this to an example playbook, e.g. https://github.com/openmicroscopy/ansible-public-omero-example or a new omero-readonly playbook.

joshmoore commented:
That all matches my concern(s) as well. I'm certainly working towards a `read-only` role, but I don't know if I'll make it in a timely fashion. Regardless, I'll look into one of the options, but this PR does have some valid tests, so I'll revert the idempotency test and get things passing then merge as "more tests"

joshmoore committed:
Disable non-idempotent schema changes

This rolls back the attempt to have a test of granting
read-only permissions. Tests of Charles inferior status
are still in place.

manics commented on playbook.yml:
```diff
@@ -10,3 +10,20 @@
+      - user: charles
+        password: charles123
+        databases: [publicdb]
+        privs: "SELECT"
```
`privs` is currently unused

joshmoore replied:
Ah, that explains something. I'm inclined to get this in (without tag) and we can come back to "privs" in https://github.com/openmicroscopy/ansible-role-postgresql/issues/5

manics reviewed (commented):
Otherwise looks fine
